File: Python/min-stack.py
class MinStack:
    # initialize your data structure here.
    def __init__(self):
        self.stack = []
        self.tracker = []

    # @param x, an integer
    # @return nothing
    def push(self, x):
        self.stack.append(x)
        if self.tracker and self.tracker[-1]<x:
            self.tracker.append(self.tracker[-1])
        else:
            self.tracker.append(x)
    # @return nothing
    def pop(self):
        self.stack.pop()
        self.tracker.pop()

    # @return an integer
    def top(self):
        return self.stack[-1]

    # @return an integer
    def getMin(self):
        return self.tracker[-1]

File: C++/min-stack.cpp
class MinStack {
private:
    std::stack<int> stk;
    std::stack<int> min_val_stk;
public:
    void push(int x) {
        stk.push(x);
        if(!min_val_stk.empty())
            min_val_stk.push(std::min(min_val_stk.top(), x));
        else
            min_val_stk.push(x);
    }

    void pop() {
        stk.pop();
        min_val_stk.pop();
    }

    int top() {
        return stk.top();
    }

    int getMin() {
        return min_val_stk.top();
    }
};

